commands/files/readerfile.go

```diff
@@ -13,10 +13,12 @@ type ReaderFile struct {
 	fullpath string
 	reader   io.ReadCloser
 	stat     os.FileInfo
+	offset   int64
+	baseInfo ExtraInfo
 }
 
-func NewReaderFile(filename, path string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
-	return &ReaderFile{filename, path, reader, stat}
+func NewReaderFile(filename, path, abspath string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
```

Comment:
I tried that and it is not going to work.  This test will fail.  The filename needs to be "planets/mars.txt" if you try to extract it it will be just "mars.txt".   You will get this:

```
expecting success: 
        mkdir mountdir/planets &&
        echo "Hello Mars!" >mountdir/planets/mars.txt &&
        echo "Hello Venus!" >mountdir/planets/venus.txt &&
        ipfs add -r mountdir/planets >actual

ok 42 - 'ipfs add -r' succeeds

expecting success: 
        PLANETS="QmWSgS32xQEcXMeqd3YPJLrNBLSdsfYCep2U7CFkyrjXwY" &&
        MARS="QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN" &&
        VENUS="QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4" &&
        echo "added $MARS planets/mars.txt" >expected &&
        echo "added $VENUS planets/venus.txt" >>expected &&
        echo "added $PLANETS planets" >>expected &&
        test_cmp expected actual

> diff -u expected actual
--- expected    2016-05-11 03:20:12.742743658 +0000
+++ actual      2016-05-11 03:20:12.466738156 +0000
@@ -1,3 +1,2 @@
-added QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN planets/mars.txt
-added QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4 planets/venus.txt
-added QmWSgS32xQEcXMeqd3YPJLrNBLSdsfYCep2U7CFkyrjXwY planets
+added QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN mars.txt
+added QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4 venus.txt

not ok 43 - 'ipfs add -r' output looks good
#
#               PLANETS="QmWSgS32xQEcXMeqd3YPJLrNBLSdsfYCep2U7CFkyrjXwY" &&
#               MARS="QmPrrHqJzto9m7SyiRzarwkqPcCSsKR2EB1AyqJfe8L8tN" &&
#               VENUS="QmU5kp3BH3B8tnWUU2Pikdb2maksBNkb92FHRr56hyghh4" &&
#               echo "added $MARS planets/mars.txt" >expected &&
#               echo "added $VENUS planets/venus.txt" >>expected &&
#               echo "added $PLANETS planets" >>expected &&
#               test_cmp expected actual
#
```

Here is a quick patch to try, it might not apply cleanly for you though.

```
diff --git a/commands/files/readerfile.go b/commands/files/readerfile.go
index 2508fe1..d7f5443 100644
--- a/commands/files/readerfile.go
+++ b/commands/files/readerfile.go
@@ -4,6 +4,7 @@ import (
        "errors"
        "io"
        "os"
+       gopath "path"
 )
 
 // ReaderFile is a implementation of File created from an `io.Reader`.
@@ -17,7 +18,11 @@ type ReaderFile struct {
        baseInfo ExtraInfo
 }
 
-func NewReaderFile(filename, path string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
+func NewReaderFile(_, path string, reader io.ReadCloser, stat os.FileInfo) *ReaderFile {
+       filename := path 
+       if path != "" {
+               filename = gopath.Base(path)
+       }
        return &ReaderFile{filename, path, reader, stat, 0, PosInfo{0, path}}
 }
```